heat/core/tiling.py

```diff
@@ -0,0 +1,756 @@
+import torch
+
+from . import dndarray
+
+__all__ = ["SquareDiagTiles"]
+
+
+def _create_cols_square_diag(arr, lshape_map, tiles_per_proc):
+    """
+    Calculates the last diagonal process, then creates a list of the number of tile columns per
+    process, then calculates the starting indices of the columns. Also returns the number of tile
+    columns.
+
+    Parameters
+    ----------
+    arr : DNDarray
+        DNDarray for which to find the tile columns for
+    lshape_map : torch.Tensor
+        the map of the local shapes (for more info see: dndarray.DNDarray.create_lshape_map())
+    tiles_per_proc : int
+        the number of divisions per process
+
+    Returns
+    -------
+    last_dia_pr : single element torch.Tensor
+        the process number which has the last diagonal tile
+    col_per_proc_list : list
+        a list of how many tile columns are on each process
+    col_inds : list
+        the starting index of each column
+    tile_columns : single element torch.Tensor
+        the number of tile columns
+    """
+    last_tile_cols = tiles_per_proc
+    last_dia_pr = torch.where(lshape_map[..., arr.split].cumsum(dim=0) >= min(arr.gshape))[0][0]
+
+    # adjust for small blocks on the last diag pr:
+    last_pr_minus1 = last_dia_pr - 1 if last_dia_pr > 0 else 0
+    rem_cols_last_pr = abs(
+        min(arr.gshape) - lshape_map[..., arr.split].cumsum(dim=0)[last_pr_minus1]
+    )
+    # this is the number of rows/columns after the last diagonal on the last diagonal pr
+
+    while 1 < rem_cols_last_pr / last_tile_cols < 2:
+        # todo: determine best value for this (prev at 2)
```

Comment:
Is this an issue?

Reply:
It is possible that it is not worth the time to change this at all as 2 is the minimum number of columns required for QR to work and datasets that we are targeting have much much more than this on each node. I think that it is very probable that any speedup which we find by modifying this value would be specific to that dataset/node configuration. That being said, I can create an issue for it as well